torch/testing/_internal/compilable_quantized_tensor.py

```diff
@@ -0,0 +1,366 @@
+        )
+
+    @classmethod
+    def implements(cls, aten_ops_or_torch_fns):  # type: ignore[no-untyped-def]
+        if not hasattr(cls, "_DISPATCH_TABLE"):
+            cls._DISPATCH_TABLE = {}  # type: ignore[attr-defined]
+
+        if not isinstance(aten_ops_or_torch_fns, (list, tuple)):
+            aten_ops_or_torch_fns = [aten_ops_or_torch_fns]
+
+        def decorator(func):  # type: ignore[no-untyped-def]
+            for op in aten_ops_or_torch_fns:
+
+                @functools.wraps(op)
+                def wrapper(f, types, args, kwargs):  # type: ignore[no-untyped-def]
+                    return func(f, types, args, kwargs)
+
+                cls._DISPATCH_TABLE[op] = wrapper  # type: ignore[attr-defined]
+
+            return func
+
+        return decorator
+
+    """There are two entry points that we can modify the behavior of a pytorch op: torch_function and torch_dispatch:
+
+    __torch_function__: will be called whenever a torch level function is called on the Tensor object,
+    for example: torch.nn.functional.linear, tensor.detach, tensor.reshape, tensor.t etc.
+
+    __torch_dispatch__: will be called in the C++ dispatcher, when an aten operator is called on the Tensor object,
+    for example: aten.mm, aten.addmm, aten.detach.default, aten.t.default etc.
+
+    We have some helper functions that can dispatch to the functions registered with MyDTypeTensor.implements,
+    but if the default implementation does not work for your use case, please feel free to customize it
+    """
+
+    @classmethod
+    def __torch_function__(cls, func, types, args, kwargs=None):  # type: ignore[no-untyped-def]
+        kwargs = {} if kwargs is None else kwargs
+
+        if hasattr(cls, "_DISPATCH_TABLE") and func in cls._DISPATCH_TABLE:
+            return cls._DISPATCH_TABLE[func](func, types, args, kwargs)
+
+        with torch._C.DisableTorchFunctionSubclass():
+            return func(*args, **kwargs)
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs):  # type: ignore[no-untyped-def]
+        kwargs = {} if kwargs is None else kwargs
+
+        if hasattr(cls, "_DISPATCH_TABLE") and func in cls._DISPATCH_TABLE:
+            return cls._DISPATCH_TABLE[func](func, types, args, kwargs)
+
+        raise NotImplementedError(
+            f"{cls.__name__} dispatch: attempting to run unimplemented operator/function: {func}"
+        )
+
+
+######################################################
+# LayoutType and Layout Tensor Subclass Registration #
+######################################################
+
+
+class PlainMyDTypeLayout(MyDTypeLayout):
+    def __new__(  # type: ignore[no-untyped-def]
+        cls,
+        int_data: torch.Tensor,
+        scale: torch.Tensor,
+        layout_type: LayoutType,
+    ):
+        kwargs = {}
+        kwargs["device"] = int_data.device
+        kwargs["layout"] = kwargs.get("layout") if kwargs.get("layout", False) else int_data.layout  # type: ignore[assignment]
```

Comment:
nit: `kwargs` is guaranteed not to contain layout here, so you can just always do `int_data.layout`?

Can you help me understand the purpose of the layout here? My feeling is this is less necessary if we hardcode a specific quantization layout in this subclass

Reply:
yeah it's not strictly needed I think, we actually override the layout concept in native Tensor for something else, see `Pakcing/Layout` section in https://github.com/pytorch/ao/issues/391

Reply:
got it, my preference would be to remove / hardcode one